src/subcommands/sign.rs

```diff
@@ -19,34 +21,63 @@ pub struct Sign {
 
     /// String of UTF-8 text
     input_data: String,
+
+    /// Encode the signature in ASN.1 format (for ECC signatures
+    /// only).
+    #[structopt(long)]
+    encode_asn1: bool,
+}
+
+#[derive(Serialize, Deserialize)]
+struct EccSignature {
+    r: IntegerAsn1,
+    s: IntegerAsn1,
 }
 
 impl Sign {
     /// Signs data.
     pub fn run(&self, basic_client: BasicClient) -> Result<()> {
-        let alg = key_attributes(&basic_client, &self.key_name)?
+        let alg = basic_client
+            .key_attributes(&self.key_name)?
             .policy
             .permitted_algorithms;
 
-        //TODO: depending on the hash in the algorithm, we need to hash input data
-
         let signature = match alg {
             Algorithm::AsymmetricSignature(alg) => {
+                info!("Hashing data...");
+                let hash = match alg.hash() {
+                    Some(SignHash::Specific(hash)) => hash_data(self.input_data.as_bytes(), hash)?,
+                    _ => {
+                        err!("Asymmetric signing algorithm ({:?}) is not supported", alg);
+                        return Err(ToolErrorKind::NotSupported.into());
+                    }
+                };
                 info!("Signing data...");
-                Ok(basic_client.psa_sign_hash(
-                    self.key_name.clone(),
-                    self.input_data.as_bytes(),
-                    alg,
-                )?)
+                let mut sig = basic_client.psa_sign_hash(self.key_name.clone(), &hash, alg)?;
+                if alg.is_ecc_alg() && self.encode_asn1 {
+                    let mut asn1_signature = vec![0; 1000];
+                    let size = picky_asn1_der::to_bytes(
```

Comment:
Can you use the `picky_asn1_der::to_vec` method to directly generate a vector of the appropriate length without having to truncate a buffer?